[#11_2] 상품 CRUD

2eve2 opened the pull request:
* 판매자 -> 상품 등록
* 상품 기본/추가 옵션 관리
* 확장성을 고려한 상품 이미지 처리

2eve2 committed:
refactor: 옵션그룹 리스트, ordering 필드 추가

2eve2 committed:
feat: docker db test 유저 생성 및 권한 부여

dding94 committed:
refactor: 아이템 도메인 수정

빌더패턴 적용

dding94 committed:
feat: ItemOptionGroup,ItemOption Repository 인터페이스 구현

dding94 committed:
feat: 아이템 옵션그룹,옵션 InMemoryRepository 구현

어댑터 패턴 적용

dding94 committed:
refactor: ItemProcessorConfig 수정

itemOption, itemOptionGroup Repository 추가

dding94 committed:
chore: import 순서 수정

dding94 committed:
refactor: 아이템 그룹옵션,옵션 저장 로직 추가

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/item/persistence/inmemory/InMemoryItemOptionGroupRepository.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class InMemoryItemOptionGroupRepository {
```
이제 데이터베이스 연동 하고 있으니 memory repository는 굳이 만드시지 않으셔도 됩니다 :)
테스트를 위해서 만드는거라면 상관 없지만요..! :)

f-lab-bright reviewed (approved):
LGTM with Comments

dding94 committed:
fix: 실수로 만든 list 삭제

dding94 committed:
refactor: json 데이터 역질렬화를 위한 기본 생성자 추가

기본생성자를 다른곳에서 사용할 수 없게 protected 선언

dding94 committed:
docs: item-api 상품등록 수정

dding94 committed:
refactor: 도메인 의존 코드 삭제

dding94 commented on src/main/java/com/flab/livecommerce/presentation/item/ItemController.java:
```diff
@@ -20,9 +19,10 @@ public ItemController(ItemManager itemManager) {
+    //todo 옵션그룹, 옵션이 JSON 변환 안되는 부분 빠르게 수정 필요
     @PostMapping
     public CommonApiResponse registerItem(@RequestBody @Valid RegisterItemRequest request) {
-        Item registeredItem = itemManager.register(request.toCommand());
-        return CommonApiResponse.success(registeredItem);
```
지난 PR 에서 엔티티를 반환해서 등록 성공시 보여주도록 코드를 작성 하셔서 저희 아키텍쳐 구조상 프레젠테이션 레이어가 도메인레이어를 직접 의존하고 있는 코드는 맞지 않다고 생각되어서 다시 지웠습니다!
보여주고 싶으신 데이터가 있다면 responseDTO 를 만들어 엔티티를 직접 노출하지 않고 변환해서 보여주는게 적합하다고 생각이 듭니다.

dding94 committed:
refactor: 옵션그룹, 옵션에서 Id를 받도록 수정, 및 커맨드 네임 변경

객체관계가 양방향관계이 되면 객체 순환참조 위험성이 있다.

dding94 committed:
refactor: 필요없는 Setter 삭제, 기본 생성자 protected 변경

protected 선언으로 의미없는 객체생성 방지

dding94 commented on src/main/java/com/flab/livecommerce/application/item/facade/ItemManager.java:
```diff
@@ -22,19 +22,19 @@ public ItemManager(
     }
 
-    public void checkProductNameDuplicated(RegisterCommand command) {
+    public void checkProductNameDuplicated(RegisterItemCommand command) {
         if (null != itemRepository.findByName(command.getName())) {
             throw new DuplicatedItemNameException("이미 존재하는 상품명입니다.");
         }
```
상품명을 중복이 안되게 하는 이유를 잘 모르겠습니다. 
간단히 아이폰 13 으로 예를 들면 판매자는 다르지만 동일한 "아이폰 13"이라고 이름을 지정하고 팔 수 있지 않을까요?

2eve2 replied:
우선 판매자만 상품 등록이 가능하기 때문에 한 판매자가 같은 상품명을 가지고 있으면 안된다고 생각했고, 다른 판매자일 경우에도 쿠팡이나 아마존 같은 플랫폼에서는 사용자의 상품 검색 편의상 완전히 같은 상품명을 사용하지 않더라고요. 

현재 판매자와 일반유저 역할이 user 도메인내에서 분리가 되어있지 않아서, api 로직에서 판매자 id를 기준으로 상품을 가져오지 않고 있습니다. 판매자 역할 추가 후에 저희가 만들어놨던 logincheck 어노테이션 체크와 itemRepository에서 판매자에 따른 상품명을 불러오도록 수정할 수 있을 것입니다.

다른 판매자인데 동일한 상품일 경우 플랫폼측에서 노출 상품명을 바꿔주는 것인지, 혹은 적절한 하나의 상품만을 노출시켜주는 것인지는 파악하기 어려워서 이 부분은 함께 논의해보면 좋을 것 같습니다.

dding94 committed:
chore: todo 삭제

dding94 committed:
feat: 복잡한 아이템 옵션그룹,옵션 저장을 위한 인터페이스 추가

dding94 committed:
feat: 아이템 등록 command 추가

dding94 committed:
feat: 아이템 옵션시리즈 인터페이스, 구현체 구현

애플리케이션 레이어 영역의 복잡한 구현로직을 인프라영역에 옮기기 위함

dding94 committed:
refactor: setter 삭제, 리스트 객체 command 로 변경

dding94 committed:
refactor: 옵션시리즈 등록 및 사용

상품 옵션 저장시 복잡한 로직 제거 및 상품 등록 커맨드는 분리 함

dding94 committed:
refactor: import 수정

dding94 committed:
docs: flyway item 추가

dding94 committed:
refactor: Item 도메인 setId 메서드 추가

dding94 committed:
feat: Item 도메인 MysqlRepository 추가

dding94 committed:
refactor: Transactional 어노테이션 추가

상품 아이템,옵션,그룹 은 하나의 단위로 저장되어야함

dding94 committed:
refactor: 옵션은 옵션그룹의 아이디를 가지도록 수정

dding94 committed:
refactor: adapter 변경

2eve2 commented on src/main/java/com/flab/livecommerce/application/item/RegisterItemProcessor.java:
```diff
@@ -1,40 +1,29 @@
+    @Transactional
+    public Item execute(RegisterItemCommand command) {
+        var item = itemRepository.save(command.toEntity());
+        itemOptionSeriesService.save(command, item);
```
이미 command.toEntity()로 등록할 item 객체를 만들어줬는데 파라미터로 command와 item 모두를 넘겨주는 이유가 뭔지 알 수 있을까요?

dding94 replied:
등록할 item 객체를 만들었지만 만든 item 에서는 OptionGroupList 가 없습니다.
아래 서비스에서 커맨드에서 받은 리스트(그룹 옵션, 옵션)를 저장하는데 만든 item 을 가져 온 이유는 그룹옵션에서 item저장을 할 때 생성된 item id가 필요하기 때문입니다.

2eve2 commented on src/main/java/com/flab/livecommerce/infrastructure/item/ItemOptionSeriesServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+
+    @Override
+    public List<ItemOptionGroup> save(RegisterItemCommand command, Item item) {
+        var itemOptionGroupList = command.getItemOptionGroup();
+
+        if (itemOptionGroupList.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        return itemOptionGroupList.stream().map(
+            requestItemOptionGroup -> {
+                var optionGroup = requestItemOptionGroup.toEntity(item);
+                var itemOptionGroup = itemOptionGroupRepository.save(optionGroup);
+
```
여기서도 itemOptionGropList는 command에서 가져오고, 밑에서는 전 단계에서 command를 item 객체로 바꾼 것에서 optionGroup을 별도로 생성해서 저장하는데 이유가 궁금합니다.

dding94 replied:
위에서 답변한걸로 설명이 가능할 것이라고 생각됩니다. toEntity 에서 item이 어떻게 사용되는지 보면 될 것 같습니다.

2eve2 commented on src/main/java/com/flab/livecommerce/infrastructure/item/ItemOptionSeriesServiceImpl.java:
```diff
@@ -0,0 +1,50 @@
+import com.flab.livecommerce.domain.item.ItemOptionGroup;
+import com.flab.livecommerce.domain.item.ItemOptionGroupRepository;
+import com.flab.livecommerce.domain.item.ItemOptionRepository;
+import com.flab.livecommerce.domain.item.ItemOptionSeriesService;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ItemOptionSeriesServiceImpl implements ItemOptionSeriesService {
+
```
ItemOptionSeriesService를 domain에 위치시킨 이유와 인터페이스로 만든 이유가 궁금합니다. 

구체적으로 기존에 만들던 방식으로 application layer에서 processor나 manager로 하지않고 service라는 명칭을 새로 쓰고, 인터페이스와 구현체로 분리한 이유가 궁금합니다. 또한 service 인터페이스를 domain layer에 위치시킨 이유도요.

dding94 replied:
이전에 작성한 코드는 너무 세세한 구현으로 한눈에 파악하기 힘든 코드였습니다. 따라서 애플리케이션 레이어 보다는 인프라스트럭쳐 레이어에 위임하는게 맞다고 생각이 들어서 따로 분리를 하였습니다. 그리고 레이어드 아키텍쳐의 흐름을 지키기 위해 도메인에 인터페이스를 두고 사용해서 dip를 통하여 구현체를 사용을 했습니다.
더 좋은 방법이 있다면 코멘트를 남겨주시면 감사하겠습니다.

dding94 committed:
feat: item search 기능 추가, 추가 작업 할점 상세보기에

null(exception) 처리, 응답시 도메인 객체 반환하지 않도록, 아이템 옵션그룹, 옵션도 반환하도록 추가 작업 필요 임시 뼈대만 구축

2eve2 commented on src/main/resources/db/migration/V1__init_ddl_.sql:
```diff
@@ -0,0 +1,43 @@
+-- user
+CREATE TABLE user
+(
+    id       bigint       not null auto_increment primary key comment 'ID',
+    email    varchar(255) not null comment '유저 email',
+    password varchar(255) not null comment '유저 패스워드',
+    nickname varchar(255) not null comment '닉네임'
+);
+
+-- item
+create table item
+(
+    id             bigint auto_increment primary key comment 'ID',
+    name           varchar(255) not null comment '상품명',
+    description    varchar(255) not null comment '상품 설명',
+    price          int(11) not null comment '상품 가격',
+    sales_price    int(11) not null comment '할인 가격',
+    stock_quantity int(11) not null comment '상품 재고',
+    model_number   int(11) not null
+);
+
```
flyway을 쓰면 DDL 형상관리가 가능하다는 장점이 있습니다. 앞으로는 지금처럼 처음에 작성한 user DDL을 삭제하고 item 관련 DDL을 추가하여 합쳐서 V1로 관리하지 말고, 다음 버전으로 새로운 table DDL만 추가하는 것이 좋을 것 같은데 어떻게 생각하시나요?

dding94 replied:
저도 이부분에 대해서 고민을 했는데 애플리케이션의 전체 테이블을 생성하는 것이 하나는 필요하다고 생각이 들어 도메인이 추가 될 때마다 v1 user, v2 item , v3 order 이름 보다는 v2 init v3 init 이라는 이름이 더 좋지 않나 생각을해서 user 를 삭제하였습니다.

f-lab-bright replied:
flyway로 삽입되는 모든 파일로 전체가 만들어지는 겁니다.
매번 같은 파일에 스키마를 추가한다면 flyway를 사용할 이유가 전혀 없습니다.
로컬에서 실행할 때도 파일이 수정되면 컨플릭이 발생해서 수동으로 해결해주거나, 전부 지운 뒤 다시 실행해줘야하니까요.
그리고 발전하는 서비스는 테이블이 계속 추가되기때문에 완전한 전체 테이블의 스키마라는건 존재할 수 없습니다.
피쳐에서 필요로하는 스키마는 새 버전의 flyway 스크립트로 작성되어서 같이 변경사항이 관리되어야 합니다.
flyway를 왜 쓰는지를 다시 고민해보시면 좋겠네요

2eve2 committed:
refactor: Repository naming 수정

- 향후 Mybatis, Jpa 사용 고려하여 구체적인 데이터 접근 기술 명시

2eve2 committed:
refactor: JdbcTemplate

- NamedParameterJdbcTemplate: INSERT 외 데이터 변경 시 sql param 바인딩 순서 모호성 제거
- BeanPropertySqlParameterSource: INSERT) 객체-> param 설정 시 불필요한 코드 제거

2eve2 committed:
Merge branch 'feature/11-product' of https://github.com/f-lab-edu/LiveCommerce into feature/11-product

 Conflicts:
	src/main/java/com/flab/livecommerce/infrastructure/item/persistence/ItemRepositoryAdapter.java
	src/main/java/com/flab/livecommerce/infrastructure/item/persistence/mysql/MysqlItemRepository.java

2eve2 committed:
feat: 상품 썸네일 이미지 기능

- Multipart 인식 문제 해결 필요
- ExceptionHandling 필요
- 이미지 파일 처리 Utils 분리 고려 중

2eve2 committed:
refactor: api 반환값

2eve2 committed:
fix: 체크스타일 개행 추가

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/item/persistence/mysql/MysqlItemRepository.java:
```diff
@@ -26,6 +29,13 @@ public Item save(Item item) {
+
+        return DataAccessUtils.singleResult(items);
+    }
+
```
사라진 코드 부분이 해당 부분입니다 확인한번 부탁드립니다. 복구는 제가 알아서 하겠습니다.

dding94 committed:
refactor: 구체적인 단어 사용

dding94 committed:
refactor: 모델넘버, 상품이름 체크 삭제

운영으로 처리 가능한 부분이기 때문

dding94 committed:
refactor: 아이템 등록, 이미지 등록 분리
